sbom/spdx/encoder.go

```diff
@@ -0,0 +1,417 @@
+package spdx
+
+import (
+	"context"
+	"fmt"
+	spdxjson "github.com/spdx/tools-golang/json"
+	"io"
+	"runtime/debug"
+	"slices"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/sbom"
+
+	"github.com/spdx/tools-golang/spdx/common"
+	v2common "github.com/spdx/tools-golang/spdx/v2/common"
+	"github.com/spdx/tools-golang/spdx/v2/v2_3"
+)
+
+type Version string
+
+const (
+	V2_3 Version = "v2.3"
+)
+
+type Format string
+
+const JSON Format = "json"
+
+// Creator describes the creator of the SPDX document that will be produced from the encoding.
+type Creator struct {
+	// Creator is the value of the [Creator] relationship.
+	Creator string
+	// CreatorType is the key of the [Creator] relationship.
+	// In accordance to the SPDX v2 spec, CreatorType should be one of "Person", "Organization", or "Tool".
+	CreatorType string
+}
+
+var _ sbom.Encoder = (*Encoder)(nil)
+
+type Encoder struct {
+	Version           Version
+	Format            Format
+	Creators          []Creator
+	DocumentName      string
+	DocumentNamespace string
+	DocumentComment   string
+}
+
+func NewDefaultEncoder(documentName, documentNamespace, documentComment string) *Encoder {
+	return &Encoder{
+		Version: V2_3,
+		Format:  JSON,
+		Creators: []Creator{
+			{
+				Creator:     fmt.Sprintf("Claircore-%s", getVersion()),
+				CreatorType: "Tool",
+			},
+		},
+		DocumentName:      documentName,
+		DocumentNamespace: documentNamespace,
+		DocumentComment:   documentComment,
+	}
+}
+
+// Encode encodes a [claircore.IndexReport] to an [io.Reader].
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, w io.Writer, ir *claircore.IndexReport) error {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return err
+	}
+
+	// TODO(blugo): support SPDX versions before 2.3
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return fmt.Errorf("unknown SPDX version: %v", e.Version)
```

Comment:
```suggestion
		return fmt.Errorf("unknown SPDX version: %w", e.Version)
```

Reply:
Could you explain this suggestion? `e.Version` isn't an error, so it's unclear to me why `%w` is preferred.

Reply:
sorry, used to this being used for errors 😄